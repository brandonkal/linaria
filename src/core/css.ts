import { StyledComponent } from '../react';

export type Interpolation =
  | string
  | number
  | CSSProperties
  | StyledComponent<any, any>;

export type CSSProperties = {
  [key: string]: string | number | CSSProperties;
};

// Override css return type for TS
type CssFn = (
  _strings: TemplateStringsArray,
  ..._exprs: Interpolation[]
) => string;

function css(_strings: TemplateStringsArray, ..._exprs: Interpolation[]): void {
  if (process.env.NODE_ENV !== 'production') {
    throw new Error(
      'Using the "css" or "injectGlobal" tag in runtime is not supported. Have you set up the Babel plugin correctly?'
    );
  }
}

export default css as CssFn;
