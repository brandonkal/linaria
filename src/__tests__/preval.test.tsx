/* eslint-disable no-template-curly-in-string */

import * as babel from '@babel/core';
import dedent from 'dedent';
import { join, resolve } from 'path';
import stripAnsi from 'strip-ansi';
import Module from '../babel/module';

beforeEach(() => Module.invalidateAll());

import serializer from './__utils__/linaria-snapshot-serializer';

expect.addSnapshotSerializer(serializer);

const babelrc = {
  babelrc: false,
  presets: [
    [require.resolve('../babel'), { displayName: true, evaluate: true }],
  ],
};

const FULL = '@brandonkal/linaria/react';
const SHORT = '../react';

const transpile = async (input: string) => {
  const replaced = input.includes(FULL);
  const codeInput = input.replace(FULL, SHORT);
  const { code, metadata } = await babel.transformAsync(codeInput, {
    ...babelrc,
    filename: join(__dirname, 'source.js'),
  });

  return {
    metadata: metadata as typeof metadata & { keepEmptyLines?: boolean },
    code: replaced ? code.replace(SHORT, FULL) : code,
  };
};

it('evaluates identifier in scope', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    const answer = 42;
    const foo = () => answer;
    const days = foo() + ' days';

    export const Title = styled.h1\`
      &:before {
        content: "${'${days}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates local expressions', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    const answer = 42;
    const foo = () => answer;

    export const Title = styled.h1\`
      &:before {
        content: "${"${foo() + ' days'}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates functions with nested identifiers', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    const objects = { key: { fontSize: 12 } };
    const foo = (k) => {
      const obj = objects[k];
      return obj;
    };

    export const Title = styled.h1\`
      ${"${foo('key')}"}
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates expressions with dependencies', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';
    import slugify from '../utils/slugify';

    export const Title = styled.h1\`
      &:before {
        content: "${"${slugify('test')}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates expressions with expressions depending on shared dependency', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';
    const slugify = require('../utils/slugify').default;

    const boo = t => slugify(t) + 'boo';
    const bar = t => slugify(t) + 'bar';

    export const Title = styled.h1\`
      &:before {
        content: "${"${boo('test') + bar('test')}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates multiple expressions with shared dependency', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';
    const slugify = require('../utils/slugify').default;

    const boo = t => slugify(t) + 'boo';
    const bar = t => slugify(t) + 'bar';

    export const Title = styled.h1\`
      &:before {
        content: "${"${boo('test')}"}"
        content: "${"${bar('test')}"}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evalutes interpolations with sequence expression', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';
    let external;

    export const Title = styled.h1\`
      color: ${'${(external, (() => "blue")())}'};
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evalutes dependencies with sequence expression', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';
    import external from './__fixtures__/sample-script.js'

    const color = (external, 'blue');

    export const Title = styled.h1\`
      color: ${'${color}'};
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates component interpolations', async () => {
  const { code, metadata } = await transpile(
    dedent`
    const { styled } = require('../react');

    export const Title = styled.h1\`
      color: red;
    \`;

    export const Paragraph = styled.p\`
      ${'${Title}'} {
        color: blue;
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('throws when interpolation evaluates to undefined', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      const { styled } = require('../react');

      let fontSize;

      export const Title = styled.h1\`
        font-size: ${'${fontSize}'};
      \`;
      `
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('throws when interpolation evaluates to null', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      const { styled } = require('../react');

      const color = null;

      export const Title = styled.h1\`
        color: ${'${color}'};
      \`;
      `
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('throws when interpolation evaluates to NaN', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      const { styled } = require('../react');

      const height = NaN;

      export const Title = styled.h1\`
        height: ${'${height}'}px;
      \`;
      `
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('handles wrapping another styled component', async () => {
  const { code, metadata } = await transpile(
    dedent`
    const { styled } = require('../react');

    const Title = styled.h1\`
      color: red;
    \`;

    export const CustomTitle = styled(Title)\`
      font-size: 24px;
      color: blue;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('inlines object styles as CSS string', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    const fill = (top = 0, left = 0, right = 0, bottom = 0) => ({
      position: 'absolute',
      top,
      right,
      bottom,
      left,
    });

    export const Title = styled.h1\`
      ${'${fill(0, 0)}'}
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('inlines array styles as CSS string', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    const fill = (top = 0, left = 0, right = 0, bottom = 0) => [
      { position: 'absolute' },
      {
        top,
        right,
        bottom,
        left,
      }
    ];

    export const Title = styled.h1\`
      ${'${fill(0, 0)}'}
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('ignores inline arrow function expressions', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    export const Title = styled.h1\`
      &:before {
        content: "${'${props => props.content}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('ignores inline vanilla function expressions', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    export const Title = styled.h1\`
      &:before {
        content: "${'${function(props) { return props.content }}'}"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('ignores external expressions', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    const generate = props => props.content;

    export const Title = styled.h1\`
      &:before {
        content: "${'${generate}'}px;"
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates complex styles with functions and nested selectors', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { css } from '../index.ts';
    export const bareIconClass = css\`\`;

    const getSizeStyles = (fs) => ({
      [\`${'&.${bareIconClass}'}\`]: {
        fontSize: fs * 1.5,
      },
    });

    export const SIZES = {
      XS: css\`${'${getSizeStyles(11)}'}\`,
    };
  `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('throws codeframe error when evaluation fails', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      import { styled } from '@brandonkal/linaria/react';

      const foo = props => { throw new Error('This will fail') };

      export const Title = styled.h1\`
        font-size: ${'${foo()}'}px;
      \`;
      `
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('handles escapes properly', async () => {
  const { code, metadata } = await babel.transformFileAsync(
    resolve(__dirname, './__fixtures__/escape-character.js'),
    babelrc
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('handles complex component', async () => {
  const { code, metadata } = await babel.transformFileAsync(
    resolve(__dirname, './__fixtures__/complex-component.js'),
    babelrc
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('generates stable class names', async () => {
  const { code, metadata } = await babel.transformFileAsync(
    resolve(__dirname, './__fixtures__/components-library.js'),
    babelrc
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('derives display name from filename', async () => {
  const { code, metadata } = await babel.transformAsync(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    export default styled.h1\`
      font-size: 14px;
    \`;
    `,
    {
      ...babelrc,
      filename: join(__dirname, 'FancyName.js'),
    }
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('derives display name from parent folder name', async () => {
  const { code, metadata } = await babel.transformAsync(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    export default styled.h1\`
      font-size: 14px;
    \`;
    `,
    {
      ...babelrc,
      filename: join(__dirname, 'FancyName/index.js'),
    }
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('throws if unable to determine a display name', async () => {
  expect.assertions(1);

  try {
    await babel.transformAsync(
      dedent`
      import { styled } from '@brandonkal/linaria/react';

      export default styled.h1\`
        font-size: 14px;
      \`;
      `,
      {
        ...babelrc,
        filename: join(__dirname, '/.js'),
      }
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('does not make identifiers arrow functions', async () => {
  const { code, metadata } = await transpile(
    dedent`
      import { styled } from '../react';
      const Input = styled.input\`\`;

      export const Page = (p => styled.div\`
        color: #fff;
        ${'${Input}'} {
          color: #241047;
        }
      \`)({})
      `
  );
  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('does not transform identifiers to arrows that contain propsName', async () => {
  const { code, metadata } = await transpile(
    dedent`
      import { styled } from '../react';
      const Classes = { p: 'hello' }

      export const Page = (p => styled.div\`
        color: #fff;
        &.${'${Classes.p}'} {
          color: #241047;
        }
      \`)({})
      `
  );
  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

function lineOf(search: string, text: string) {
  const lines = text.split('\n');
  return lines.findIndex(line => line.includes(search)) + 1;
}

it('creates valid CSS with multiline interpolation', async () => {
  const source = dedent`
  import { styled } from '../react';

  function getNumber() {
    return 10
  }

  export const Page = (p => styled.div\`
    font-size: ${'${getNumber() +\n\
      2}'}px;
    background: black;
  \`)({})
  `;
  const { code, metadata } = await transpile(source);
  expect(code).toMatchSnapshot();
  expect(lineOf('background:', (metadata as any).linaria.cssText)).toBe(
    lineOf('background:', source)
  );
  metadata.keepEmptyLines = true;
  expect(metadata).toMatchSnapshot();
});

it('does not transform identifiers to arrows that contain propsName 2', async () => {
  const { code, metadata } = await transpile(
    dedent`
      import { styled } from '../react';
      const Classes = { p: 'hello' }
      const ap = Classes

      export const Page = (p => styled.div\`
        color: #fff;
        &.${'${Classes["p"]}'} {
          color: #241047;
        }
        &.${'${ap["p"]}'} {
          color: #241047;
        }
      \`)({})
      `
  );
  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('does not strip instanbul coverage sequences', async () => {
  const { code, metadata } = await babel.transformAsync(
    dedent`
    import { styled } from '@brandonkal/linaria/react';

    const a = 42;

    export const Title = styled.h1\`
      height: ${'${a}'}px;
    \`;
    `,
    {
      ...babelrc,
      cwd: '/home/user/project',
      filename: 'file.js',
      plugins: [
        [
          // eslint-disable-next-line import/no-extraneous-dependencies
          require('babel-plugin-istanbul').default,
          { cwd: '/home/user/project' },
        ],
      ],
    }
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});
