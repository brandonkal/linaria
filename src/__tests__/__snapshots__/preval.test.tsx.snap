// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`derives display name from filename 1`] = `
"import { styled } from '@brandonkal/linaria/react';
export default
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"FancyName0\\",
  class: \\"FancyName0_f1yjazsq\\"
});"
`;

exports[`derives display name from filename 2`] = `

CSS:

.FancyName0_f1yjazsq {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`derives display name from parent folder name 1`] = `
"import { styled } from '@brandonkal/linaria/react';
export default
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"FancyName0\\",
  class: \\"FancyName0_f1e8u5dq\\"
});"
`;

exports[`derives display name from parent folder name 2`] = `

CSS:

.FancyName0_f1e8u5dq {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`does not make identifiers arrow functions 1`] = `
"import { styled } from '../react';
const Input =
/*#__PURE__*/
styled(\\"input\\")({
  name: \\"Input\\",
  class: \\"Input_i1xha7dm\\"
});
export const Page =
/*#__PURE__*/
styled(\\"div\\")({
  name: \\"Page\\",
  class: \\"Page_p1rsdnkv\\"
});"
`;

exports[`does not make identifiers arrow functions 2`] = `

CSS:

.Input_i1xha7dm {}

.Page_p1rsdnkv {
color: #fff;
.Input_i1xha7dm {
  color: #241047;
}
}

Dependencies: ../react

`;

exports[`does not strip instanbul coverage sequences 1`] = `
"var cov_2dr9r1nt95 = function () {
  var path = \\"/home/user/project/file.js\\";
  var hash = \\"23f56bbaeb2ebcd213f41374d3b2af1bce287bd3\\";
  var global = new Function(\\"return this\\")();
  var gcv = \\"__coverage__\\";
  var coverageData = {
    path: \\"/home/user/project/file.js\\",
    statementMap: {
      \\"0\\": {
        start: {
          line: 3,
          column: 10
        },
        end: {
          line: 3,
          column: 12
        }
      },
      \\"1\\": {
        start: {
          line: 5,
          column: 21
        },
        end: {
          line: 7,
          column: 1
        }
      }
    },
    fnMap: {},
    branchMap: {},
    s: {
      \\"0\\": 0,
      \\"1\\": 0
    },
    f: {},
    b: {},
    _coverageSchema: \\"43e27e138ebf9cfc5966b082cf9a028302ed4184\\",
    hash: \\"23f56bbaeb2ebcd213f41374d3b2af1bce287bd3\\"
  };
  var coverage = global[gcv] || (global[gcv] = {});

  if (coverage[path] && coverage[path].hash === hash) {
    return coverage[path];
  }

  return coverage[path] = coverageData;
}();

import { styled } from '@brandonkal/linaria/react';
const a = (cov_2dr9r1nt95.s[0]++, 42);
export const Title = (cov_2dr9r1nt95.s[1]++,
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_tow9xsn\\"
}));"
`;

exports[`does not strip instanbul coverage sequences 2`] = `

CSS:

.Title_tow9xsn {
  height: 42px;
}

Dependencies: NA

`;

exports[`does not transform identifiers to arrows that contain propsName 1`] = `
"import { styled } from '../react';
const Classes = {
  p: 'hello'
};
export const Page =
/*#__PURE__*/
styled(\\"div\\")({
  name: \\"Page\\",
  class: \\"Page_p1xha7dm\\"
});"
`;

exports[`does not transform identifiers to arrows that contain propsName 2`] = `

CSS:

.Page_p1xha7dm {
color: #fff;
&hello {
  color: #241047;
}
}

Dependencies: ../react

`;

exports[`does not transform identifiers to arrows that contain propsName 2 1`] = `
"import { styled } from '../react';
const Classes = {
  p: 'hello'
};
const ap = Classes;
export const Page =
/*#__PURE__*/
styled(\\"div\\")({
  name: \\"Page\\",
  class: \\"Page_p1xha7dm\\"
});"
`;

exports[`does not transform identifiers to arrows that contain propsName 2 2`] = `

CSS:

.Page_p1xha7dm {
color: #fff;
&hello {
  color: #241047;
}
&hello {
  color: #241047;
}
}

Dependencies: ../react

`;

exports[`evaluates complex styles with functions and nested selectors 1`] = `
"import { css } from '../index.ts';
export const bareIconClass = \\"bareIconClass_b1xha7dm\\";

const getSizeStyles = fs => ({
  [\`&.\${bareIconClass}\`]: {
    fontSize: fs * 1.5
  }
});

export const SIZES = {
  XS: \\"XS_x1rsdnkv\\"
};"
`;

exports[`evaluates complex styles with functions and nested selectors 2`] = `

CSS:

.bareIconClass_b1xha7dm {}
.XS_x1rsdnkv {&.bareIconClass_b1xha7dm { font-size: 16.5px; }}

Dependencies: ../index.ts

`;

exports[`evaluates component interpolations 1`] = `
"const {
  styled
} = require('../react');

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});
export const Paragraph =
/*#__PURE__*/
styled(\\"p\\")({
  name: \\"Paragraph\\",
  class: \\"Paragraph_p1rsdnkv\\"
});"
`;

exports[`evaluates component interpolations 2`] = `

CSS:

.Title_t1xha7dm {
  color: red;
}

.Paragraph_p1rsdnkv {
  .Title_t1xha7dm {
    color: blue;
  }
}

Dependencies: ../react

`;

exports[`evaluates expressions with dependencies 1`] = `
"import { styled } from '@brandonkal/linaria/react';
import slugify from '../utils/slugify';
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates expressions with dependencies 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "6og6jy"
  }
}

Dependencies: @babel/runtime/helpers/interopRequireDefault, ../react, ../utils/slugify

`;

exports[`evaluates expressions with expressions depending on shared dependency 1`] = `
"import { styled } from '@brandonkal/linaria/react';

const slugify = require('../utils/slugify').default;

const boo = t => slugify(t) + 'boo';

const bar = t => slugify(t) + 'bar';

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates expressions with expressions depending on shared dependency 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "6og6jyboo6og6jybar"
  }
}

Dependencies: ../react, ../utils/slugify

`;

exports[`evaluates functions with nested identifiers 1`] = `
"import { styled } from '@brandonkal/linaria/react';
const objects = {
  key: {
    fontSize: 12
  }
};

const foo = k => {
  const obj = objects[k];
  return obj;
};

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates functions with nested identifiers 2`] = `

CSS:

.Title_t1xha7dm {
  font-size: 12px;
}

Dependencies: ../react

`;

exports[`evaluates identifier in scope 1`] = `
"import { styled } from '@brandonkal/linaria/react';
const answer = 42;

const foo = () => answer;

const days = foo() + ' days';
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates identifier in scope 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "42 days"
  }
}

Dependencies: ../react

`;

exports[`evaluates local expressions 1`] = `
"import { styled } from '@brandonkal/linaria/react';
const answer = 42;

const foo = () => answer;

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates local expressions 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "42 days"
  }
}

Dependencies: ../react

`;

exports[`evaluates multiple expressions with shared dependency 1`] = `
"import { styled } from '@brandonkal/linaria/react';

const slugify = require('../utils/slugify').default;

const boo = t => slugify(t) + 'boo';

const bar = t => slugify(t) + 'bar';

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evaluates multiple expressions with shared dependency 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "6og6jyboo"
    content: "6og6jybar"
  }
}

Dependencies: ../react, ../utils/slugify

`;

exports[`evalutes dependencies with sequence expression 1`] = `
"import { styled } from '@brandonkal/linaria/react';
import external from './__fixtures__/sample-script.js';
const color = (external, 'blue');
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evalutes dependencies with sequence expression 2`] = `

CSS:

.Title_t1xha7dm {
  color: blue;
}

Dependencies: NA

`;

exports[`evalutes interpolations with sequence expression 1`] = `
"import { styled } from '@brandonkal/linaria/react';
let external;
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`evalutes interpolations with sequence expression 2`] = `

CSS:

.Title_t1xha7dm {
  color: blue;
}

Dependencies: ../react

`;

exports[`generates stable class names 1`] = `
"import { styled } from '../../react';
export const T1 =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"T1\\",
  class: \\"T1_tq14ovd\\"
});
export const T2 =
/*#__PURE__*/
styled(\\"h2\\")({
  name: \\"T2\\",
  class: \\"T2_t1mjbcp5\\"
});
export const T3 =
/*#__PURE__*/
styled(\\"h3\\")({
  name: \\"T3\\",
  class: \\"T3_t1ij2n2c\\"
});"
`;

exports[`generates stable class names 2`] = `

CSS:

.T1_tq14ovd {
  background: #111;
}
.T2_t1mjbcp5 {
  background: #222;
}
.T3_t1ij2n2c {
  .T2_t1mjbcp5 {
    background: #333;
  }
}

Dependencies: ../../react

`;

exports[`handles complex component 1`] = `
"import { styled } from '../../react';
const objects = {
  font: {
    fontSize: 12
  },
  box: {
    border: '1px solid red'
  }
};

const foo = k => {
  const {
    [k]: obj
  } = objects;
  return obj;
};

objects.font.fontWeight = 'bold';
export const whiteColor = '#fff';
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t16ua7kf\\"
});"
`;

exports[`handles complex component 2`] = `

CSS:

.Title_t16ua7kf {
  font-size: 12px; font-weight: bold;
  border: 1px solid red;
}

Dependencies: ../../react

`;

exports[`handles escapes properly 1`] = `
"import { styled } from '../../react';
const selectors = ['a', 'b'];
export const Block =
/*#__PURE__*/
styled(\\"div\\")({
  name: \\"Block\\",
  class: \\"Block_b3cgn7e\\"
});"
`;

exports[`handles escapes properly 2`] = `

CSS:

.Block_b3cgn7e {
  a { content: "\\u000A"; } b { content: "\\u000A"; };
}

Dependencies: ../../react

`;

exports[`handles wrapping another styled component 1`] = `
"const {
  styled
} = require('../react');

const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});
export const CustomTitle =
/*#__PURE__*/
styled(Title)({
  name: \\"CustomTitle\\",
  class: \\"CustomTitle_c1rsdnkv\\"
});"
`;

exports[`handles wrapping another styled component 2`] = `

CSS:

.Title_t1xha7dm {
  color: red;
}

.CustomTitle_c1rsdnkv.Title_t1xha7dm {
  font-size: 24px;
  color: blue;
}

Dependencies: ../react

`;

exports[`ignores external expressions 1`] = `
"import { styled } from '@brandonkal/linaria/react';

const generate = props => props.content;

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\",
  vars: {
    \\"t1xha7dm-0\\": [generate]
  }
});"
`;

exports[`ignores external expressions 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "var(--t1xha7dm-0)"
  }
}

Dependencies: ../react

`;

exports[`ignores inline arrow function expressions 1`] = `
"import { styled } from '@brandonkal/linaria/react';
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\",
  vars: {
    \\"t1xha7dm-0\\": [props => props.content]
  }
});"
`;

exports[`ignores inline arrow function expressions 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "var(--t1xha7dm-0)"
  }
}

Dependencies: NA

`;

exports[`ignores inline vanilla function expressions 1`] = `
"import { styled } from '@brandonkal/linaria/react';
export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\",
  vars: {
    \\"t1xha7dm-0\\": [function (props) {
      return props.content;
    }]
  }
});"
`;

exports[`ignores inline vanilla function expressions 2`] = `

CSS:

.Title_t1xha7dm {
  &:before {
    content: "var(--t1xha7dm-0)"
  }
}

Dependencies: NA

`;

exports[`inlines array styles as CSS string 1`] = `
"import { styled } from '@brandonkal/linaria/react';

const fill = (top = 0, left = 0, right = 0, bottom = 0) => [{
  position: 'absolute'
}, {
  top,
  right,
  bottom,
  left
}];

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`inlines array styles as CSS string 2`] = `

CSS:

.Title_t1xha7dm {
  position: absolute; top: 0; right: 0; bottom: 0; left: 0;
}

Dependencies: ../react

`;

exports[`inlines object styles as CSS string 1`] = `
"import { styled } from '@brandonkal/linaria/react';

const fill = (top = 0, left = 0, right = 0, bottom = 0) => ({
  position: 'absolute',
  top,
  right,
  bottom,
  left
});

export const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_t1xha7dm\\"
});"
`;

exports[`inlines object styles as CSS string 2`] = `

CSS:

.Title_t1xha7dm {
  position: absolute; top: 0; right: 0; bottom: 0; left: 0;
}

Dependencies: ../react

`;

exports[`throws codeframe error when evaluation fails 1`] = `
"<<DIRNAME>>/source.js: An error occurred when evaluating the expression: This will fail.
Make sure you are not using a browser or Node specific API.
  4 | 
  5 | export const Title = styled.h1\`
> 6 |   font-size: \${foo()}px;
    |                ^
  7 | \`;"
`;

exports[`throws if couldn't determine a display name 1`] = `
"<<DIRNAME>>/.js: Couldn't determine a name for the component. Ensure that it's either:
- Assigned to a variable
- Is an object property
- Is a prop in a JSX element

  1 | import { styled } from '@brandonkal/linaria/react';
  2 | 
> 3 | export default styled.h1\`
    |                ^
  4 |   font-size: 14px;
  5 | \`;"
`;

exports[`throws when interpolation evaluates to NaN 1`] = `
"<<DIRNAME>>/source.js: 
The expression evaluated to 'NaN', which is probably a mistake.
If you want it to be inserted into CSS, explicitly cast or transform the value to a string, e.g. - 'String(height)'.
  4 | 
  5 | export const Title = styled.h1\`
> 6 |   height: \${height}px;
    |             ^
  7 | \`;"
`;

exports[`throws when interpolation evaluates to null 1`] = `
"<<DIRNAME>>/source.js: 
The expression evaluated to 'null', which is probably a mistake.
If you want it to be inserted into CSS, explicitly cast or transform the value to a string, e.g. - 'String(color)'.
  4 | 
  5 | export const Title = styled.h1\`
> 6 |   color: \${color};
    |            ^
  7 | \`;"
`;

exports[`throws when interpolation evaluates to undefined 1`] = `
"<<DIRNAME>>/source.js: 
The expression evaluated to 'undefined', which is probably a mistake.
If you want it to be inserted into CSS, explicitly cast or transform the value to a string, e.g. - 'String(fontSize)'.
  4 | 
  5 | export const Title = styled.h1\`
> 6 |   font-size: \${fontSize};
    |                ^
  7 | \`;"
`;
